core/src/main/java/io/grpc/internal/SpiffeUtil2.java

```diff
@@ -0,0 +1,229 @@
+      if (domainNode == null || domainNode.size() == 0) {
+        trustBundleMap.put(trustDomainName, Collections.emptyList());
+        continue;
+      }
+      Long sequenceNumber = JsonUtil.getNumberAsLong(domainNode, "sequence_number");
+      sequenceNumbers.put(trustDomainName, sequenceNumber);
+      List<Map<String, ?>> keysNode = JsonUtil.getListOfObjects(domainNode, "keys");
+      if (keysNode == null || keysNode.size() == 0) {
+        trustBundleMap.put(trustDomainName, Collections.emptyList());
+        continue;
+      }
+      trustBundleMap.put(trustDomainName, extractCerts(keysNode, trustDomainName));
+    }
+    return new SpiffeBundle(sequenceNumbers, trustBundleMap);
+  }
+
+  private static Map<String, ?> readTrustDomainsFromFile(String filePath) throws IOException {
+    Path path = Paths.get(checkNotNull(filePath, "trustBundleFile"));
+    String json = new String(Files.readAllBytes(path), StandardCharsets.UTF_8);
+    Object jsonObject = JsonParser.parseNoDuplicates(json);
+    if (!(jsonObject instanceof Map)) {
+      throw new IllegalArgumentException(
+          "SPIFFE Trust Bundle should be a JSON object. Found: "
+              + (jsonObject == null ? null : jsonObject.getClass()));
+    }
+    @SuppressWarnings("unchecked")
+    Map<String, ?> root = (Map<String, ?>)jsonObject;
+    Map<String, ?> trustDomainsNode = JsonUtil.getObject(root, "trust_domains");
+    checkArgument(checkNotNull(trustDomainsNode,
+            "Mandatory trust_domains element is missing").size() > 0,
+        "Mandatory trust_domains element is missing");
+    return trustDomainsNode;
+  }
+
+  private static List<X509Certificate> extractCerts(List<Map<String, ?>> keysNode,
+      String trustDomainName) {
+    List<X509Certificate> result = new ArrayList<>();
+    for (Map<String, ?> keyNode : keysNode) {
+      String kid = JsonUtil.getString(keyNode, "kid");
+      if (kid != null && !kid.equals("")) {
+        log.log(Level.SEVERE, String.format("'kid' parameter must not be set but value '%s' "
+                + "found. Skipping certificate loading for trust domain '%s'.", kid,
+            trustDomainName));
+        break;
+      }
+      String use = JsonUtil.getString(keyNode, "use");
+      if (use == null || !use.equals(USE_PARAMETER_VALUE)) {
+        log.log(Level.SEVERE, String.format("'use' parameter must be '%s' but '%s' found. "
+                + "Skipping certificate loading for trust domain '%s'.", USE_PARAMETER_VALUE, use,
+            trustDomainName));
+        break;
+      }
+      String rawCert = JsonUtil.getString(keyNode, "x5c");
+      if (rawCert == null) {
+        break;
+      }
+      InputStream stream = new ByteArrayInputStream(rawCert.getBytes(StandardCharsets.UTF_8));
+      try {
+        Collection<? extends Certificate> certs = CertificateFactory.getInstance("X509")
+            .generateCertificates(stream);
+        if (certs.size() > 0) {
```

Comment:
Could this be > 1? If yes, should we take those into account? What does the spec say about this case?

Reply:
Yes, it can be >1 (it's not an exception). IIUC we can simply ignore it - https://datatracker.ietf.org/doc/html/rfc7517#section-4.7
WDYT?

Reply:
Taking back my previous comment - it must be [exactly one](https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md#61-publishing-spiffe-bundle-elements)
Changed accordingly